[DOCS] Add MetricProviders Conceptual Content (#8175)
great-expectations/great_expectations: docs/docusaurus/docs/conceptual_guides/metricproviders.md

- Conceptually grouping code for Metrics calculation across multiple backends.
- Incremental development. You can develop Metrics for each Execution Engine, one at a time.
- Sharing Metrics across Expectations.
- Generating Metrics that can be inferred from specific types of Expectation automatically.

## Assumed knowledge

To get the most out of the information provided here, you should have an understanding of:

- Expectations
- Metrics
- The Metric registry
- ExecutionEngines

## When MetricProviders are required

A minimum of one supporting Metric is required by every Expectation. For example, `expect_column_mean_to_be_between` relies on a Metric that calculates the mean of a column. Often, an Expectation requires multiple Metrics. For example, the following Metrics are required in the `expect_column_values_to_be_in_set` Expectation:

- `column_values.in_set.unexpected_count`
- `column_values.in_set.unexpected_rows`
- `column_values.in_set.unexpected_values`
- `column_values.in_set.unexpected_value_counts`

> To allow Expectations to work with multiple backends, methods for calculating Metrics need to be implemented for each ExecutionEngine. For example the `aggregate_fn` utility that is used with SQL and Spark requires the `column_values.in_set.unexpected_count.aggregate_fn` Metric. 

Reviewer: I think this would be a stronger example if it were explaining the different methods that are used to calculate the same metric in different ExecutionEngines.  Something along the lines of: "For example, the `some_metric` in `ExpectSomethingToBeTrue` is implemented with the `some_spark_method` utility for the Spark ExecutionEngines, but the pandas ExecutionEngine calculates the metric using `something_different`.

Reviewer: You'll probably need feedback from someone with more engineering knowledge to put together that kind of an example, but I think it would be more meaningful given the topic sentence here.

Author: @austiezr could you provide the information necessary to update the text as Rachel has recommended, or could you provide the name of a contact who can provide the necessary information?

Reviewer 2: This `column_mean` [Metric](https://github.com/great-expectations/great_expectations/blob/develop/great_expectations/expectations/metrics/column_aggregate_metrics/column_mean.py) is a good example -- `_pandas` is implemented by calling the Pandas-native `.mean()` method on the column, whereas `_spark` is implemented with a Spark-native mean function & `_sqlalchemy` is implemented with a SQLAlchemy generic function.

Author: The text was updated as recommended by @austiezr.

> To allow Expectations to work with multiple backends, methods for calculating Metrics need to be implemented for each ExecutionEngine. For example the `aggregate_fn` utility that is used with SQL and Spark requires the `column_values.in_set.unexpected_count.aggregate_fn` Metric. 

Author: ```suggestion
To allow Expectations to work with multiple backends, methods for calculating Metrics need to be implemented for each ExecutionEngine. For example, pandas is implemented by calling the built-in pandas `.mean()` method on the column, Spark is implemented with a built-in Spark `mean` function, and SQLAlchemy is implemented with a SQLAlchemy generic function.
```

> Metrics can help you incorporate conditional statements in Expectations that support conditional evaluations. For example, `column_values.in_set.condition`.

Reviewer: We have the name of a metric here, but how is it used to incorporate conditional statements in Expectations?  Maybe expand this to be something like: "For example, `column_values.in_set.condition` can be set when metrics are calculated and then used as a flag in your Custom Expectation to determine branching behavior."  (Assuming that's an accurate explanation of the example.  I'd double check with someone more familiar with Custom Expectations.)

> Metrics such as `column_values.in_set.unexpected_index_list` and `column_values.in_set.unexpected_index_query` can help you calculate the truthiness of your data.

Reviewer: Again, I think expanding this statement with an explanation of "how" would make for a stronger claim.

Author: Discussed with @austiezr and we agreed that this isn't necessary here.

        ColumnPairMapMetricProvider
        MulticolumnMapExpectation
        MulticolumnMapMetricProvider
```

If you’re not sure which MetricProvider you should use with an Expectation, you can usually infer the correct MetricProvider class name from the Expectation class name. For example:

| Expectation class                 | MetricProvider class                    |
| --------------------------------- | --------------------------------------- |
| Expectation                       | MetricProvider                          |
| BatchExpectation                  | TableMetricProvider                     |
| ColumnAggregateExpectation        | ColumnAggregateMetricProvider           |
| ColumnMapExpectation              | ColumnMapMetricProvider                 |
| RegexBasedColumnMapExpectation    | RegexColumnMapMetricProvider (built in) |
| SetBasedColumnMapExpectation      | SetColumnMapMetricProvider (built in)   |
| ColumnPairMapExpectation          | ColumnPairMapMetricProvider             |
| MulticolumnMapExpectation         | MulticolumnMapMetricProvider            |
| QueryExpectation                  | QueryMetricProvider                     |

> Sometimes, the MetricProvider class is created directly from the Expectation class, so you don’t need to specify a MetricProvider or methods when declaring a new Expectation. For example, the RegexColumnMapMetricProvider and the SetColumnMapMetricProvider classes.

Reviewer: The example feels like a fragment.  Maybe have it be something like "For example, the RegexBasedColumnMapExpectation automatically implements the RegexColumnMapMetricProvider, and the SetBasedColumnMapExpectation automatically implements the SetColumnMapMetricProvider."

It would also be ideal to provide a link to a reference of which Expectations automatically implement a MetricProvider, but I'm not certain we have one currently.  It might be worth suggesting an update be made to Expectation class docstrings so that they indicate if a MetricProvider is automatically implemented with them, and then link to the Expectations section of the API docs in a future PR.

Author: @Rachel-Reverie I've implemented the text edit you recommended. Who could help with the Expectation class docstrings  you have suggested?

## Define Metrics with a MetricProvider

> The API for MetricProvider classes is unusual. MetricProvider classes are never intended to be instantiated, and they don’t have inputs or outputs in the normal sense of method arguments and return values. Instead, the inputs for MetricProvider classes are methods for calculating the Metric on different backend applications. Each method must be decorated with an appropriate decorator. On `new`, the class registers the methods as part of the Metrics registry so that they can be invoked as Metrics. The registered Metrics are the only output from MetricProviders.

Reviewer: ```suggestion
The API for MetricProvider classes is unusual. MetricProvider classes are never intended to be instantiated, and they don’t have inputs or outputs in the normal sense of method arguments and return values. Instead, the inputs for MetricProvider classes are methods for calculating the Metric on different backend applications. Each method must be decorated with an appropriate decorator. On `new`, the MetricProvider class registers the decorated methods as part of the Metrics registry so that they can be invoked to calculate Metrics. The registered methods are the only output from MetricProviders.
```

:::note
Decorators invoked on `new` can make maintainability challenging. GX intends to address this shortcoming in future releases.
:::

> A typical MetricProvider class has three methods that correspond to the three primary ExecutionEngines used by GX (pandas, SQL and SparkDF). Each of the three methods are marked with an appropriate decorator.

Reviewer: ```suggestion
A typical MetricProvider class has three methods that correspond to the three primary ExecutionEngines used by GX (pandas, SQL, and SparkDF). Each of the three methods are marked with an appropriate decorator.
```

Each MetricProvider class supports different decorators. Every MetricProvider class can use `@metric_value` and `@metric_partial`. For the MetricProvider, QueryMetricProvider, and TableMetricProvider classes, these are the only supported Metric decorators.

The following table lists the MetricProvider subclasses and their associated Metrics.

| Subclass                          | Supported Metrics                                              |
| --------------------------------- | ---------------------------------------------------------------|
| ColumnAggregateMetricProvider     | @column_aggregate_value, @column_aggregate_partial             |
| ColumnMapMetricProvider           | @column_condition_partial, @column_function_partial            |
| ColumnPairMapMetricProvider       | @column_pair_condition_partial                                 |
| MulticolumnMapMetricProvider      | @multicolumn_condition_partial, @multicolumn_function_partial  |

## Metric decorator naming conventions

The following Metric decorators differ in terms of the types of inputs and outputs they accept:

- A metric_value returns
- A metric_partials

> - condition_partials versus function_partials

Reviewer: This section seems incomplete.
